docs/dev/datastream_execution_mode.md

```diff
@@ -0,0 +1,242 @@
+
+## Configuring BATCH execution mode
+
+The execution mode can be configured via the `execution.runtime-mode` setting.
+There are three possible values:
+
+ - `STREAMING`: The classic DataStream execution mode
+ - `BATCH`: Batch-style execution on the DataStream API
+ - `AUTOMATIC`: Let the system decide based on the boundedness of the sources
+
+ This can be configured either in the `flink-conf.yaml`, via command line
+ parameters of `bin/flink run ...`, or programmatically when
+ creating/configuring the `StreamExecutionEnvironment`.
+
+ Here's how you can configure the execution mode via the command line:
+
+ ```bash
+ $ bin/flink run -Dexecution.runtime-mode=BATCH examples/streaming/WordCount.jar
+ ```
+
+ This example shows how you can configure the execution mode in code:
+
+ ```java
+Configuration config = new Configuration();
+config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
+ ```
+
+## Execution Behavior
+
+This section provides an overview of the execution behavior of BATCH execution
+mode and contrasts it with STREAMING execution mode. For more details, please
+refer to the FLIPs that introduced this feature:
+[FLIP-134](https://cwiki.apache.org/confluence/x/4i94CQ) and
+[FLIP-140](https://cwiki.apache.org/confluence/x/kDh4CQ). As well as the
+documentation about [task scheduling (TODO)](<TODO>).
+
+### Task Scheduling And Network Shuffle
+
+Flink jobs consist of different operations that are connected together in a
+dataflow graph. The system decides how to schedule the execution of these
+operations on different processes/machines (TaskManager) and how data is
+shuffled (sent) betwixt them.
+
+Multiple operations/operators can be chained together using a feature called
+[chaining]({% link dev/stream/operators/index.md
+%}#task-chaining-and-resource-groups). A group of one or multiple (chained)
+operators that Flink considers as a unit of scheduling is called a _task_.
+Often the term _subtask_ is used to refer to the individual instances of tasks
+that are running in parallel on multiple TaskManagers but we will only use the
+term _task_ here.
+
+Task scheduling and the network shuffle work different for BATCH execution mode
+and STREAMING execution mode. Mostly due to the fact that we know our input
+data is bounded in BATCH execution mode, which allows Flink to use more
+efficient data structures and algorithms.
+
+We will use this example to explain the differences in task scheduling and
+network transfer:
+
+```java
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+DataStreamSource<String> source = env.fromElements(...);
+
+source.name("source")
+	.map(...).name("map1")
+	.map(...).name("map2")
+	.rebalance()
+	.map(...).name("map3")
+	.map(...).name("map4")
+	.keyBy((value) -> value)
+	.map(...).name("map5")
+	.map(...).name("map6")
+	.sinkTo(...).name("sink");
```

Comment:
A diagram of an execution graph for this topology would be helpful.

Reply:
I was afraid someone would suggest this, but yes, I assumed that I would have to do this... 😅🙈

Reply:
I added a figure.